src/SleighAsm.h

```diff
@@ -6,67 +6,274 @@
+
+		if(type != UNIQUE)
+			throw LowlevelError("Only unique vars will be added into unordered set.");
 
-        ~AssemblySlg() { if(str) free(str); }
+		return self.offset;
+	}
+
+	bool is_unique() const { return type == UNIQUE; }
+
+	bool is_const() const { return type == CONST; }
+
+	bool is_ram() const { return type == RAM; }
+
+	bool is_reg() const { return type == REGISTER; }
+};
+
+ostream &operator<<(ostream &s, const PcodeOperand &arg);
+
+typedef OpCode PcodeOpType;
+
+struct Pcodeop
+{
+	PcodeOpType type;
+
+	PcodeOperand *output = nullptr;
+	PcodeOperand *input0 = nullptr;
+	PcodeOperand *input1 = nullptr;
+	/* input2 for STORE will use output to save memory space */
+
+	Pcodeop(PcodeOpType opc, PcodeOperand *in0, PcodeOperand *in1, PcodeOperand *out):
+	    type(opc), input0(in0), input1(in1), output(out)
+	{
+	}
+
+	void fini()
+	{
+		if(output)
+			delete output;
+		if(input0)
+			delete input0;
+		if(input1)
+			delete input1;
+	}
 };
 
+ostream &operator<<(ostream &s, const Pcodeop &op);
+
+class PcodeSlg : public PcodeEmit
+{
+private:
+	PcodeOperand *parse_vardata(VarnodeData &data)
+	{
+		AddrSpace *space = data.space;
+		PcodeOperand *operand = nullptr;
+		if(space->getName() == "register")
+		{
+			operand = new PcodeOperand(
+			    space->getTrans()->getRegisterName(data.space, data.offset, data.size), data.size);
+			operand->type = PcodeOperand::REGISTER;
+		}
+		else if(space->getName() == "ram")
+		{
+			operand = new PcodeOperand(data.offset, data.size);
+			operand->type = PcodeOperand::RAM;
+		}
+		else if(space->getName() == "const")
+		{
+			// space.cc's ConstantSpace::printRaw()
+			operand = new PcodeOperand(data.offset);
+			operand->type = PcodeOperand::CONST;
+			operand->size = data.size; // To aviod ctor's signature collide with RAM's
+		}
+		else if(space->getName() == "unique")
+		{
+			operand = new PcodeOperand(data.offset, data.size);
+			operand->type = PcodeOperand::UNIQUE;
+		}
+		else if(space->getName() == "DATA")
+		{
+			operand = new PcodeOperand(data.offset, data.size);
+			operand->type = PcodeOperand::RAM;
+		}
+		else
+			throw LowlevelError("Unsupported AddrSpace type appear.");
+		return operand;
+	}
+
+public:
+	std::vector<Pcodeop> pcodes;
+
+	void dump(const Address &addr, OpCode opc, VarnodeData *outvar, VarnodeData *vars,
+	          int4 isize) override
+	{
+		PcodeOperand *out = nullptr, *in0 = nullptr, *in1 = nullptr;
+
+		switch(isize)
+		{
+			case 3: out = parse_vardata(vars[2]); // Only for STORE
+			case 2: in1 = parse_vardata(vars[1]);
+			case 1: in0 = parse_vardata(vars[0]);
+			case 0: break;
+			default: throw LowlevelError("Unexpexted isize in PcodeSlg::dump()");
+		}
+
+		if(outvar)
+			out = parse_vardata(*outvar);
+
+		pcodes.push_back(Pcodeop(opc, in0, in1, out));
+	}
+
+	~PcodeSlg()
+	{
+		while(!pcodes.empty())
+		{
+			pcodes.back().fini();
+			pcodes.pop_back();
+		}
+	}
+};
+
+struct R2Reg
+{
+	std::string name;
+	size_t size;
+	size_t offset;
```

Comment:
This and `size` should be `ut64` I think.

Reply:
Yes